File: iosApp/iosApp/screens/speaker/SpeakerViewModel.swift
//
//  SpeakerDetailViewModel.swift
//  iosApp
//
//  Created by GERARD on 17/02/2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SharedDi
import KMPNativeCoroutinesAsync

enum SpeakerUiState {
    case loading
    case success(SpeakerUi)
    case failure
}

@MainActor
class SpeakerViewModel: ObservableObject {
    private let repository: AgendaRepository = RepositoryHelper().agendaRepository
    private let alarmScheduler: AlarmScheduler = AlarmScheduler()
    let speakerId: String

    init(speakerId: String) {
        self.speakerId = speakerId
    }

    @Published var uiState: SpeakerUiState = SpeakerUiState.loading
    
    private var speakerTask: Task<(), Never>?
    
    func fetchSpeakerDetails() {
        speakerTask = Task {
            do {
                let stream = asyncSequence(for: repository.speaker(speakerId: self.speakerId))
                for try await speaker in stream {
                    self.uiState = .success(speaker)
                }
            } catch {
                self.uiState = .failure
            }
        }
    }

    func markAsFavorite(talkItem: TalkItemUi) async {
        await self.alarmScheduler.schedule(talkItem: talkItem)
    }
    
    func stop() {
        speakerTask?.cancel()
    }
}

File: android-data/src/main/java/org/gdglille/devfest/android/data/viewmodels/SpeakerViewModel.kt
package org.gdglille.devfest.android.data.viewmodels

import android.annotation.SuppressLint
import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.google.firebase.crashlytics.ktx.crashlytics
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.gdglille.devfest.android.data.AlarmScheduler
import org.gdglille.devfest.models.SpeakerUi
import org.gdglille.devfest.models.TalkItemUi
import org.gdglille.devfest.repositories.AgendaRepository

sealed class SpeakerUiState {
    data class Loading(val speaker: SpeakerUi) : SpeakerUiState()
    data class Success(val speaker: SpeakerUi) : SpeakerUiState()
    data class Failure(val throwable: Throwable) : SpeakerUiState()
}

class SpeakerViewModel(
    private val speakerId: String,
    private val repository: AgendaRepository,
    private val alarmScheduler: AlarmScheduler
) : ViewModel() {
    private val _uiState = MutableStateFlow<SpeakerUiState>(SpeakerUiState.Loading(SpeakerUi.fake))
    val uiState: StateFlow<SpeakerUiState> = _uiState

    init {
        viewModelScope.launch {
            try {
                repository.speaker(speakerId).collect {
                    _uiState.value = SpeakerUiState.Success(it)
                }
            } catch (error: Throwable) {
                Firebase.crashlytics.recordException(error)
                _uiState.value = SpeakerUiState.Failure(error)
            }
        }
    }

    @SuppressLint("UnspecifiedImmutableFlag")
    fun markAsFavorite(context: Context, talkItem: TalkItemUi) = viewModelScope.launch {
        alarmScheduler.schedule(context, talkItem)
    }

    object Factory {
        fun create(speakerId: String, repository: AgendaRepository, alarmScheduler: AlarmScheduler) =
            object : ViewModelProvider.Factory {
                override fun <T : ViewModel> create(modelClass: Class<T>): T {
                    @Suppress("UNCHECKED_CAST")
                    return SpeakerViewModel(speakerId, repository, alarmScheduler) as T
                }
            }
    }
}
